syntax = "proto3";
package kaldi_serve;

service KaldiServe {
  // Performs synchronous speech recognition: receive results
  // after all audio has been sent and processed.  
  rpc Recognize(RecognizeRequest) returns (RecognizeResponse) {}

  // Performs asynchronous speech recognition: receive results
  // with long running operation
  // rpc LongRunningRecognize(LongRunningRecognizeRequest) 
  //     returns (LongRunningRecognizeResponse) {}

  // Performs bidirectional streaming speech recognition: receive results 
  // while sending audio.
  // rpc StreamingRecognize(stream StreamingRecognizeRequest)
  //     returns (stream StreamingRecognizeResponse) {}
}

message RecognizeRequest {
  RecognitionConfig config = 1;
  RecognitionAudio audio = 2;
  string uuid = 3;
}

// message LongRunningRecognizeRequest {
//   RecognitionConfig config = 1;
//   RecognitionAudio audio = 2;
//   string uuid = 3;
// }

message RecognizeResponse {
  repeated SpeechRecognitionResult results = 1;
}

// message LongRunningRecognizeResponse {
//   repeated SpeechRecognitionResult results = 1;
// }

// Provides information to the recognizer that specifies how to process the request
message RecognitionConfig {
  enum AudioEncoding {
    ENCODING_UNSPECIFIED = 0;
    LINEAR16 = 1;
    FLAC = 2;
    // MULAW = 3;
    // AMR = 4;
    // AMR_WB = 5;
    // OGG_OPUS = 6;
    // SPEEX_WITH_HEADER_BYTE = 7;
  }

  AudioEncoding encoding = 1;
  int32 sample_rate_hertz = 2; // Valid values are: 8000-48000.
  string language_code = 3;
  int32 max_alternatives = 4;
  bool punctuation = 5;
  repeated SpeechContext speech_contexts = 6;
  int32 audio_channel_count = 7;
  bool enable_word_time_offsets = 8;
  // RecognitionMetadata metadata = 9;  
  
  string model = 10;
}

// Either `content` or `uri` must be supplied.
message RecognitionAudio {
  oneof audio_source {
    bytes content = 1;
    string uri = 2;
  }
}

message SpeechRecognitionResult {
  repeated SpeechRecognitionAlternative alternatives = 1;
}

message SpeechRecognitionAlternative {
  string transcript = 1;
  float confidence = 2;
}

message SpeechContext {
  repeated string phrases = 1;
  string type = 2;
}
